EndTime: make public interface use std::chrono

lrusak opened the pull request:
This is a follow up to #18783 that I made a year ago. This pushes the `std::chrono` usage into the public interface.

IMO this help make the code a lot more readable when the values are assigned units.

I decided to template the entire `EndTime` class, not so much for the duration but for the underlying type. With these changes you can make your timer use `double` or `float` easily. The default template uses `std::chrono::milliseconds` which can be seen by the update from `EndTime` -> `EndTime<>`. When declaring your type you could use something like `EndTime<std::chrono::duration<double, std::nano>>` to have a double, nanosecond value.

When calling `GetTimeLeft` it returns the templated duration (ie. `std::chrono::milliseconds` by default)

One thing to note that you will see in the diff is that an `EndTime` declaration may use `std::chrono::milliseconds` but when calling `Set()` I use a different unit (ie. seconds, or hours). This is perfectly valid usage as it will implicitly cast to the type. You can cast down in units (s -> ms) but not up in units (ms -> s) as the value may truncate (ie 1234ms -> 1s).
ref: https://en.cppreference.com/w/cpp/chrono/duration

I've also included a commit to use a static start time for a couple methods.

I've also renamed `MillisLeft()` to `GetTimeLeft()` for obvious reasons.

I've done more work to further expand `std::chrono` into some of the changes in this PR instead of casting to `std::chrono::milliseconds` but I didn't include it in this PR as it's already quite large and will be in follow up PR's

lrusak committed:
EndTime: make public interface use std::chrono

lrusak committed:
EndTime: rename method to GetTimeLeft

garbear commented on xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAE.cpp:
```diff
@@ -352,7 +354,7 @@ void CActiveAE::StateMachine(int signal, Protocol *port, Message *msg)
           return;
         case CActiveAEControlProtocol::KEEPCONFIG:
-          m_extKeepConfig = *(unsigned int*)msg->data;
+          m_extKeepConfig = std::chrono::milliseconds(*(unsigned int*)msg->data);
```
Opportunity to use c++ cast

lrusak replied:
Good catch. Thanks!

garbear commented on xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESink.cpp:
```diff
@@ -540,7 +540,7 @@ void CActiveAESink::StateMachine(int signal, Protocol *port, Message *msg)
           m_sink = nullptr;
           m_state = S_TOP_CONFIGURED_SUSPEND;
-          m_extTimeout = 10000;
+          m_extTimeout = 10000ms;
```
Does C++17 have a separator that lets you do something like 10_000ms?

garbear replied:
otherwise, representing as 10s might be more clear

lrusak replied:
Thanks! I've updated all 5 digit values to use the next size unit.

lrusak commented:
jenkins build this please

ksooo commented on xbmc/cores/AudioEngine/Engines/ActiveAE/ActiveAESink.cpp:
```diff
@@ -320,15 +320,15 @@ void CActiveAESink::StateMachine(int signal, Protocol *port, Message *msg)
 
         case CSinkControlProtocol::SETSILENCETIMEOUT:
-          m_silenceTimeOut = *(int*)msg->data;
+          m_silenceTimeOut = std::chrono::milliseconds(*(int*)msg->data);
```
c++ style cast, please.

lrusak replied:
fixed, thanks

ksooo commented on xbmc/threads/test/TestEndTime.cpp:
```diff
@@ -10,41 +10,43 @@
+  EXPECT_LT(0ms, endTime.GetTimeLeft());
 
-  std::this_thread::sleep_for(std::chrono::milliseconds(100));
+  std::this_thread::sleep_for(101ms);
```
100ms, not 101 !

lrusak replied:
This was deliberate as the test doesn't pass otherwise

```
$ ./kodi-test --gtest_filter="TestEndTime*" 
Note: Google Test filter = TestEndTime*
[==========] Running 2 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 2 tests from TestEndTime
[ RUN      ] TestEndTime.DefaultConstructor
[       OK ] TestEndTime.DefaultConstructor (100 ms)
[ RUN      ] TestEndTime.ExplicitConstructor
/home/lukas/Documents/git/xbmc/xbmc/threads/test/TestEndTime.cpp:28: Failure
Value of: endTime.IsTimePast()
  Actual: false
Expected: true
[  FAILED  ] TestEndTime.ExplicitConstructor (101 ms)
[----------] 2 tests from TestEndTime (201 ms total)

[----------] Global test environment tear-down
[==========] 2 tests from 1 test suite ran. (250 ms total)
[  PASSED  ] 1 test.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] TestEndTime.ExplicitConstructor

 1 FAILED TEST
```

fritsch replied:
```
    if (now <= (m_startTime + m_totalWaitTime))
      return false;
 ```
 Do you mean btw. "<" - that's why the 100ms test does not work above I assume. If you start at 0 and duration is 100 ms, then when the clock shows 100, you still return false, which is not correct.

lrusak replied:
That's a good observation but I believe it has the desired result. If the duration is `100ms` and you wait `100ms` that shouldn't be considered past, you need to wait `101ms` to be considered past.

I believe the difference comes from using the `std::chrono::time_point_cast<T>`. This changes the `time_point` to be in the templated type. The default `time_point` is using `std::chrono::nanoseconds`. This makes it so that the actual checked value is truncated to the units of measurement.

This is demonstrated in this sample programs. https://godbolt.org/z/xYq6nf3PE

with `using unit = std::chrono::nanoseconds;`
```
future: 30344022422788ns
now: 30344022500756ns
30344022500756ns > 30344022422788ns
time_waited: 100077968ns
100077968ns > 100ms
```

with `using unit = std::chrono::milliseconds;`
```
future: 30672482ms
now: 30672482ms
30672482ms == 30672482ms
time_waited: 100ms
100ms == 100ms
```

So I guess the discussion should be that if it's desired to wait `100ms` should it be considered with the smallest time point possible or with the unit desired to wait? I'll probably need to think about this for a bit to come to a conclusion however I am thinking that any unit of time passed the desired wait time should be considered past in which case the current behaviour in this PR is wrong. I'm unsure how this was working before as everything was in `milliseconds` with an `int`.

fritsch replied:
Before this PR:
```
return ((std::chrono::steady_clock::now() - m_startTime) >= m_totalWaitTime);
```

Old: now - start >= duration
New: now <= (start + duration) <=> now - start <= duration for negative case => now - start > duration for the positive case. So, yes behaviour changed from >= to >.

That's why I posted the comment above.

fritsch replied:
And to add: to get it "exact" in ms range - according to Shannon Nicquist, clock precission at least double as exact as ms is needed :-) ...

lrusak replied:
> Before this PR:
> 
> ```
> 
> return ((std::chrono::steady_clock::now() - m_startTime) >= m_totalWaitTime);
> 
> ```
> 
> 
> 
> Old: now - start >= duration
> 
> New: now <= (start + duration) <=> now - start <= duration for negative case => now - start > duration for the positive case. So, yes behaviour changed from >= to >.
> 
> 
> 
> That's why I posted the comment above. 

I'm sorry I don't follow. The sample program shows the same behavior with both methods and only that the precision of the clock influences the behavior. What am I missing?

fritsch replied:
From reading the code, the new code changed the behaviour for the "==" case. Old code: IsTimePast() returned true for this case, new code returns false. Or did I read this wrong?

lrusak replied:
> From reading the code, the new code changed the behaviour for the "==" case. Old code: IsTimePast() returned true for this case, new code returns false. Or did I read this wrong?

Yes you are correct.

ksooo reviewed (commented):
Minors only.

lrusak committed:
EndTime: make GetStartTime return a time_point

lrusak committed:
TestEndTime: add new test and template CommonTests

lrusak commented:
I've updated this with a couple changes:
 - The `time_point` is always the maximum value (int64_t and nanoseconds). Any conversions are done from here (if needed).
 - The original behaviour of calculating `IsTimePast()` has been restored (thanks @fritsch for catching that).

I also added an extra test to use the templated method.

lrusak commented:
jenkins build this please

lrusak commented:
jenkins build this please

fritsch commented on xbmc/cores/AudioEngine/Sinks/AESinkPULSE.cpp:
```diff
@@ -1089,20 +1089,21 @@ unsigned int CAESinkPULSE::AddPackets(uint8_t **data, unsigned int frames, unsig
   void *buffer = data[0]+offset*m_format.m_frameSize;
-  unsigned int wait_time = static_cast<unsigned int>(1000.0 * m_BufferSize / m_BytesPerSecond);
-  m_extTimer.Set(wait_time);
+  auto wait_time =
+      std::chrono::duration<double>(static_cast<double>(m_BufferSize) / m_BytesPerSecond);
```
The default duration is now assumed as seconds, so  the 1000 as factor with that assumption can go away. Though it's not good practice as the standard does not explicitely say that this is "seconds" which is why the unit should be specified always: https://stackoverflow.com/questions/39196817/what-is-the-default-unit-of-time-of-cs-chrono <- here is the same question.

Might be finally the reason why Keep Audio alive is not working anymore properly, as depending on OS std::chrono::duraction is something else.
